Template: Logic Module Payment Client [SC-914]

fabianschu opened the pull request.

fabianschu committed:
chore: basic boilerplate

fabianschu committed:
chore: adds basic functionality to template (involves roles)

fabianschu committed:
chore: cleanup

fabianschu committed:
fix: re-add comment

fabianschu committed:
chore: fmt

Zitzak reviewed (changes requested):
Mostly inverter Standard related comment ser

0xNuggan commented on src/templates/tests/unit/LM_PC_Template_v1.t.sol:
```diff
@@ -0,0 +1,176 @@
+        );
+        paymentClient.deposit(invalidAmount);
+    }
+
```
I would add a test for "happy path works" here

0xNuggan commented on src/templates/tests/unit/LM_PC_Template_v1.t.sol:
```diff
@@ -0,0 +1,176 @@
+        │       └── Then a payment order should be created and processed
+        └── Given caller doesn't have DEPOSIT_ADMIN_ROLE 
+            └── When trying to process a deposit
+                └── Then it should revert with CallerNotAuthorized (not done here)
```
Maybe instead of "not done here", say that this case is tested in the RoleAuthorizer or sth like that? Although a quick fail test would probably be optimal

0xNuggan commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+import {ERC165Upgradeable} from
+    "@oz-up/utils/introspection/ERC165Upgradeable.sol";
+
+// System under Test (SuT)
```
Maybe remove this since it isn't a test file?

marvinkruse commented on src/templates/modules/ILM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,74 @@
+/**
+ * @title   Inverter Template Logic Module Payment Client
+ *
+ * @notice  A template logic module payment client that handles deposits and payment processing.
```
Line length (in general in the heading section)

marvinkruse commented on src/templates/modules/ILM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,74 @@
+    /// @param  amount_ The amount of tokens to deposit.
+    function deposit(uint amount_) external;
+
+    /// @notice Process a specific deposit by calling processPayments on the payment processor
```
line length and missing dot at the end

marvinkruse commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+/**
+ * @title   Inverter Template Logic Module Payment Client
+ *
+ * @notice  A template logic module payment client that handles deposits and payment processing.
```
Line length

marvinkruse commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+            || super.supportsInterface(interfaceId_);
+    }
+
+    //--------------------------------------------------------------------------
```
Missing the space between the `// ` and `----` part

marvinkruse commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+    {
+        uint amount = _depositedAmounts[user_];
+
+        // Clear the deposit amount before processing
```
```suggestion
        // Clear the deposit amount before processing.
```

marvinkruse commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+        // Clear the deposit amount before processing
+        _depositedAmounts[user_] = 0;
+
+        // Create and add payment order
```
```suggestion
        // Create and add payment order.
```

marvinkruse commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+
+        _addPaymentOrder(order);
+
+        // Process the payment
```
Even if the comment is a bit pointless (it doesn't add any information, but i see the point of wanting to show that we add comments where it makes sense).
```suggestion
        // Process the payment.
```

marvinkruse commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+        return _depositedAmounts[user_];
+    }
+
+    //--------------------------------------------------------------------------
```
Missing space again.

marvinkruse commented on src/templates/tests/unit/LM_PC_Template_v1_Exposed.sol:
```diff
@@ -0,0 +1,14 @@
+
+// Access Mock of the PP_Template_v1 contract for Testing.
+contract LM_PC_Template_v1_Exposed is LM_PC_Template_v1 {
+    // Use the `exposed_` prefix for functions to expose internal contract for
```
```suggestion
    // Use the `exposed_` prefix for functions to expose internal functions for
```

fabianschu commented:
😂

fabianschu committed:
fix: comment

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

fabianschu committed:
fix: missing fullstop

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

fabianschu committed:
fix: missing ful

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

fabianschu committed:
fix: missing fullstop

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

fabianschu committed:
chore: formatting & role assignment

fabianschu committed:
fix: line length

fabianschu committed:
chore: adds more test cases

fabianschu committed:
fix: tree diagrams

fabianschu committed:
chore: adds tests for getDepositedAmount

fabianschu committed:
chore: more input validation

fabianschu committed:
chore: improves gherkin trees

fabianschu committed:
chore: natspec

fabianschu committed:
chore: natspec

fabianschu committed:
chore: natspec

fabianschu committed:
chore: adds sectup docs

fabianschu committed:
chore: fmt

Zitzak commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,205 @@
+            || super.supportsInterface(interfaceId_);
+    }
+
+    //--------------------------------------------------------------------------
```
```suggestion
    // -------------------------------------------------------------------------
```

Zitzak replied:
This one was still in. Going to create suggestions for all the Inverter standard stuff, so it's easy to commit

Zitzak commented on src/templates/modules/ILM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,88 @@
+    function deposit(uint amount_) external;
+
+    /// @notice Processes a user's deposit.
+    /// @param user_ The address of the user whose deposit to process.
```
```suggestion
    /// @param  user_ The address of the user whose deposit to process.
```

Zitzak commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,205 @@
+    /// @notice The maximum deposit amount.
+    uint internal constant _maxDepositAmount = 100 ether;
+
+    //--------------------------------------------------------------------------
```
```suggestion
    // --------------------------------------------------------------------------
```

Zitzak commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,205 @@
+    /// @notice The role that allows processing deposits
+    bytes32 public constant DEPOSIT_ADMIN_ROLE = "DEPOSIT_ADMIN";
+
+    /// @notice    Mapping of user addresses to their deposited token amounts.
```
```suggestion
    /// @notice Mapping of user addresses to their deposited token amounts.
```

Zitzak commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,205 @@
+    /// @notice    Mapping of user addresses to their deposited token amounts.
+    mapping(address user => uint amount) internal _depositedAmounts;
+
+    /// @notice    Payment token.
```
```suggestion
    /// @notice Payment token.
```

Zitzak commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,205 @@
+    /// @notice    Payment token.
+    IERC20 internal _paymentToken;
+
+    /// @notice    Storage gap for future upgrades.
```
```suggestion
    /// @notice Storage gap for future upgrades.
```

Zitzak commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+        external
+        onlyModuleRole(DEPOSIT_ADMIN_ROLE)
+    {
+        uint amount = _depositedAmounts[user_];
```
Hasn't been resolved. Could you give a quick comment in case you disagreed to mine :)

Zitzak reviewed (changes requested):
Few comments seemed to be still un-addressed. I also added "suggestions" to all Inverter Standard related comments, which might be a good compromise

fabianschu committed:
chore: more  missing whitespaces

fabianschu committed:
chore: constants in capslock

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>

fabianschu committed:
chore: gherkin

fabianschu committed:
chore: getters and docs

fabianschu committed:
chore: adds flags

fabianschu commented on src/templates/modules/LM_PC_Template_v1.sol:
```diff
@@ -0,0 +1,190 @@
+            originChainId: block.chainid,
+            targetChainId: block.chainid,
+            flags: 0,
+            data: new bytes32[](0)
```
good idea

fabianschu committed:
chore: adds docs for flags

Zitzak committed:
refactor: LM_PC_Template_v1

What has been done?
- Add getters for MAX_DEPOSIT_AMOUNT & DEPOSIT_ADMIN_ROLE
- Make flags internal
- Move getters to begining file

Zitzak committed:
refactor: LM_PC_Template_v1.t.sol

What has been done?
- Add tests for getters
- Move getter tests to external tests in files
- Renamed gherkin
- Make testing contract admin by default
- Add fuzzing where possible

Zitzak reviewed (approved):
LGTM ser
